pricelist_by_category_qty/models/sale.py

```diff
@@ -0,0 +1,108 @@
+#    the Free Software Foundation, either version 3 of the License, or
+#    (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU General Public License for more details.
+#
+#    You should have received a copy of the GNU General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#################################################################################
+
+from odoo import api, models
+
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    def _get_search_vals_for_quantity(self, line):
+        return [
+            ('order_id', '=', line.order_id.id),
+            ('state', '!=', 'cancel'),
+            ('product_id.categ_id', '=', line.product_id.categ_id.id),
+        ]
+
+    def _get_quantity_to_compute(self, line):
+        quantity = line.product_uom_qty
+        if line.product_id:
+            order_line_obj = self.env['sale.order.line']
+            search_domain = self._get_search_vals_for_quantity(line)
+            line_ids = order_line_obj.search(search_domain)
+            quantity = sum([x.product_uom_qty for x in line_ids])
+        return quantity
+
+    def _get_price_of_line(self, product_id,
+                           qty, partner_id, pricelist_id):
+        return self.env['product.pricelist']._price_rule_get_multi(
+            pricelist=pricelist_id,
+            products_by_qty_by_partner=[(product_id, qty, partner_id)])
+
+    def _check_if_edit(self, res, product_id):
+        if res.get(product_id):
+            price, item_id = res.get(product_id, False)
+            if item_id:
+                item = self.env['product.pricelist.item'].browse(item_id)
+                if item.categ_id:
+                    return True
+        return False
+
+    def _get_child_pricelist(self, res):
+        if res.get('item_id'):
+            item_id = res.get('item_id', False)
+            if item_id:
+                item = self.env['product.pricelist.item'].browse(item_id)
+                if item.base_pricelist_id:
+                    return item.base_pricelist_id
+        return False
+
+    def compute_global_discount(self):
+        for sale in self:
+            if sale.state in ['draft', 'sent']:
+                sale.order_line.compute_global_discount()
+        return True
+
+
+class SaleOrderLine(models.Model):
+    _inherit = 'sale.order.line'
+
+    @api.multi
+    def compute_global_discount(self):
+        for line in self:
+            sale = line.order_id.with_context(date=line.order_id.date_order)
+            partner_id = sale.partner_id.id
+            if line.product_id:
+                pricelist_id = sale.pricelist_id
+                product_id = line.product_id
+                qty = sale._get_quantity_to_compute(line)
```

Comment:
In _get_quantity_to_compute, prefer a real multi implementation as you will do as many search as lines numbers.

So, you could do it before the loop here under and build a dict of quantities per line.

Reply:
I understand your request for improvement but is it really necessary? I took over an existing module and I do not develop often enough to support this refactoring.

Reply:
I share your opinion @rousseldenis but for a first contribution this module could be merged as it with an 'Alpha' value in `development_status` key in \_\_manifest\_\_.py with a ROADMAP.rst in readme directory to indicate that module should be benefit of refactoring to work with good performance.
On my side, if I decide to use it, I'll convince my customer to pay me to optimize it.
@mayjo79 you also may add a CONTRIBUTORS.rst file (in readme) to mention you as contributors.
Thanks a lot

Reply:
Ok, so include that in ROADMAP

Reply:
All is done as expected